Player

rlcheng opened the pull request.

rlcheng committed:
- Refactor ui-console/get-position to return player position as integer instead of string.
- Add human-player namespace.
- Add human-player/move let human player make a move and return board after placing the move.

rlcheng committed:
- Refactor ui-console to split out universal ui functionality into the ui namespace. This
  refactor is to setup abstractions so that it's easier to add in other ui types in the
  future.
- Refactor human-player due to changes on ui-console.

dcmoore commented on spec/tic_tac_toe/human_player_spec.clj:
```diff
@@ -0,0 +1,25 @@
+  (:require [speclj.core :refer :all]
+            [tic-tac-toe.human-player :refer :all]))
+
+(def X "X")
```
Define this in your board ns

dcmoore commented on spec/tic_tac_toe/human_player_spec.clj:
```diff
@@ -0,0 +1,25 @@
+(def X "X")
+(def empty-board-3x3
+  [" " " " " "
+   " " " " " "
```
use the space var you already defined in your board ns

dcmoore commented on spec/tic_tac_toe/human_player_spec.clj:
```diff
@@ -0,0 +1,25 @@
+  (read-line))
+
+(defn mock-output [message]
+  (println message))
```
it's not a mock if you're actually reading in/out values

rlcheng committed:
- Refactor def O and def X. Move from specs to board to remove duplication through specs.
- Use def space from board in all specs to remove magic numbers.
- Rename ui-spec/mock-prompt to test-prompt.
- Refactor ui-spec/mock-output to be a true mock.
- Add empty line spacing for consistency across code.

Akcbryant commented on spec/tic_tac_toe/board_spec.clj:
```diff
@@ -3,24 +3,23 @@
+   space space space space
+   space space space space
+   space space space space])
+
```
This added space makes it more readable, but I would say consistency is most important.

rlcheng replied:
updated

Akcbryant commented on src/tic_tac_toe/ui.clj:
```diff
@@ -0,0 +1,29 @@
+
+(defn init [fn-prompt fn-output]
+  (def prompt fn-prompt)
+  (def output fn-output))
```
If somebody happens to not call this the rest of the functions would break, is this a problem?

Akcbryant commented on src/tic_tac_toe/board.clj:
```diff
@@ -1,6 +1,10 @@
+
+(def O "O")
+
 (def space " ")
```
Don't necessarily need the space here.

rlcheng replied:
updated

Akcbryant commented:
👍

malinatran commented on src/tic_tac_toe/human_player.clj:
```diff
@@ -0,0 +1,7 @@
+  (:require [tic-tac-toe.board :as board]
+            [tic-tac-toe.ui :as ui]))
+
+(defn move [board marker prompt output]
```
I would probably make this a verb since it's a method name (e.g. `makeMove` or `getMove`).

rlcheng replied:
updated

malinatran commented on spec/tic_tac_toe/ui_spec.clj:
```diff
@@ -0,0 +1,57 @@
+  (read-line))
+
+(defn mock-output [message]
+  (do message))
```
I believe you would use `do` if you're evaluating multiple expressions. Here you can just return `message`.

rlcheng replied:
updated

malinatran commented:
👍 LGTM.
